engine/runtime/src/main/java/org/enso/interpreter/node/controlflow/permission/PermissionGuardNode.java

```diff
@@ -0,0 +1,53 @@
+package org.enso.interpreter.node.controlflow.permission;
+
+import com.oracle.truffle.api.frame.FrameUtil;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.profiles.BranchProfile;
+import org.enso.interpreter.node.ExpressionNode;
+import org.enso.interpreter.runtime.Context;
+import org.enso.interpreter.runtime.data.text.Text;
+import org.enso.interpreter.runtime.error.PanicException;
+import org.enso.interpreter.runtime.state.State;
+
+public class PermissionGuardNode extends ExpressionNode {
+  private @Child ExpressionNode body;
+  private final boolean checkInput;
+  private final boolean checkOutput;
+  private final BranchProfile inputDisallowed = BranchProfile.create();
+  private final BranchProfile outputDisallowed = BranchProfile.create();
+
+  public PermissionGuardNode(ExpressionNode body, boolean checkInput, boolean checkOutput) {
+    this.body = body;
+    this.checkInput = checkInput;
+    this.checkOutput = checkOutput;
+  }
+
+  @Override
+  public Object executeGeneric(VirtualFrame frame) {
+    State state = (State) FrameUtil.getObjectSafe(frame, getStateFrameSlot());
```

Comment:
While evaluating I found :
```
  protected FrameSlot getStateFrameSlot() {
    if (stateFrameSlot == null) {
      CompilerDirectives.transferToInterpreterAndInvalidate();
      stateFrameSlot = ((EnsoRootNode) getRootNode()).getStateFrameSlot();
    }
    return stateFrameSlot;
  }
```
I'd like to ask? Does it make sense to have per node local `stateFrameSlot`? It might be enough to:
```
  protected FrameSlot getStateFrameSlot() {
    return((EnsoRootNode) getRootNode()).getStateFrameSlot();
  }
```
Should have the same performance (on _fast path_) because `getRootNode()` is a compilation constant.